File: backend/ai/reasoner.py
import json
import logging
from datetime import datetime, timedelta, date

import anthropic

from config import get_settings
from db.models import AIAnalysis
from db.database import SessionLocal
from ai.schemas import AIAnalysisResult, ScreeningResult, MarketOutlook
from ai.prompts import (
    build_analysis_prompt,
    build_screening_prompt,
    build_market_outlook_prompt,
)

logger = logging.getLogger(__name__)


def _strip_code_fences(text: str) -> str:
    """Remove markdown code fences (```json ... ```) from AI responses."""
    text = text.strip()
    if text.startswith("```"):
        lines = text.split("\n")
        lines = [l for l in lines if not l.strip().startswith("```")]
        text = "\n".join(lines).strip()
    return text


class AIReasoner:
    """Thin orchestration layer between deterministic analytics and the Claude API."""

    def __init__(self) -> None:
        self.settings = get_settings()
        self.client = anthropic.Anthropic(api_key=self.settings.anthropic_api_key)

    # ------------------------------------------------------------------
    # Public methods
    # ------------------------------------------------------------------

    async def analyze_ticker(
        self,
        ticker: str,
        technical_data: dict,
        fundamental_data: dict,
        regime_data: dict,
        portfolio_exposure: dict,
        deep: bool = False,
    ) -> AIAnalysisResult:
        """Run scenario analysis for a single ticker.

        Uses claude-haiku-4-5-20251001 for standard speed and
        claude-sonnet-4-6 for deep analysis.
        """
        analysis_type = "deep" if deep else "standard"
        model = (
            self.settings.ai_model_deep if deep else self.settings.ai_model_screening
        )

        # --- Check cache ---
        session = SessionLocal()
        try:
            cached = self._check_cache(ticker, analysis_type, session)
            if cached is not None:
                logger.info("Cache hit for %s (%s)", ticker, analysis_type)
                return self._db_row_to_result(cached)

            # --- Build prompt & call Claude ---
            system_prompt, user_prompt = build_analysis_prompt(
                ticker, technical_data, fundamental_data, regime_data, portfolio_exposure,
            )

            message = self.client.messages.create(
                model=model,
                max_tokens=2000,
                system=system_prompt,
                messages=[{"role": "user", "content": user_prompt}],
            )
            response_text = _strip_code_fences(message.content[0].text)

            # --- Parse & validate ---
            result = AIAnalysisResult.model_validate_json(response_text)

            # --- Persist ---
            self._save_to_db(ticker, result, analysis_type, model, session)

            return result

        except Exception:
            session.rollback()
            raise
        finally:
            session.close()

    async def screen_tickers(
        self,
        tickers_data: list[dict],
    ) -> list[ScreeningResult]:
        """Batch-screen a list of tickers and return action recommendations."""
        model = self.settings.ai_model_screening

        system_prompt, user_prompt = build_screening_prompt(tickers_data)

        message = self.client.messages.create(
            model=model,
            max_tokens=2000,
            system=system_prompt,
            messages=[{"role": "user", "content": user_prompt}],
        )
        response_text = _strip_code_fences(message.content[0].text)

        raw_list = json.loads(response_text)
        return [ScreeningResult.model_validate(item) for item in raw_list]

    async def market_outlook(
        self,
        regime_data: dict,
        macro_data: dict,
        sector_data: dict,
    ) -> MarketOutlook:
        """Generate a market-level outlook."""
        model = self.settings.ai_model_screening

        system_prompt, user_prompt = build_market_outlook_prompt(
            regime_data, macro_data, sector_data,
        )

        message = self.client.messages.create(
            model=model,
            max_tokens=2000,
            system=system_prompt,
            messages=[{"role": "user", "content": user_prompt}],
        )
        response_text = _strip_code_fences(message.content[0].text)

        return MarketOutlook.model_validate_json(response_text)

    # ------------------------------------------------------------------
    # Cache helpers
    # ------------------------------------------------------------------

    def _check_cache(self, ticker: str, analysis_type: str, session) -> AIAnalysis | None:
        """Return the most recent cached row if it is within ai_cache_hours."""
        cutoff = datetime.utcnow() - timedelta(hours=self.settings.ai_cache_hours)
        row = (
            session.query(AIAnalysis)
            .filter(
                AIAnalysis.ticker == ticker,
                AIAnalysis.analysis_type == analysis_type,
                AIAnalysis.created_at >= cutoff,
            )
            .order_by(AIAnalysis.created_at.desc())
            .first()
        )
        return row

    def _save_to_db(
        self,
        ticker: str,
        result: AIAnalysisResult,
        analysis_type: str,
        model_used: str,
        session,
    ) -> None:
        """Persist an AIAnalysisResult to the AIAnalysis table."""
        row = AIAnalysis(
            ticker=ticker,
            date=date.today(),
            analysis_type=analysis_type,
            model_used=model_used,
            bull_probability=result.bull_case.probability,
            bull_target=result.bull_case.target,
            bull_thesis=result.bull_case.thesis,
            base_probability=result.base_case.probability,
            base_target=result.base_case.target,
            base_thesis=result.base_case.thesis,
            bear_probability=result.bear_case.probability,
            bear_target=result.bear_case.target,
            bear_thesis=result.bear_case.thesis,
            risk_factors_json=result.risk_factors,
            drawdown_estimate=result.max_drawdown_estimate,
            position_size_suggestion=result.position_size_pct,
            confidence_score=result.confidence,
            timeframe=result.timeframe,
            created_at=datetime.utcnow(),
        )
        session.add(row)
        session.commit()
        logger.info("Saved %s analysis for %s to DB", analysis_type, ticker)

    # ------------------------------------------------------------------
    # Conversion helper
    # ------------------------------------------------------------------

    @staticmethod
    def _db_row_to_result(row: AIAnalysis) -> AIAnalysisResult:
        """Reconstruct an AIAnalysisResult from a cached DB row."""
        from ai.schemas import ScenarioCase

        return AIAnalysisResult(
            bull_case=ScenarioCase(
                probability=row.bull_probability,
                target=row.bull_target,
                thesis=row.bull_thesis,
            ),
            base_case=ScenarioCase(
                probability=row.base_probability,
                target=row.base_target,
                thesis=row.base_thesis,
            ),
            bear_case=ScenarioCase(
                probability=row.bear_probability,
                target=row.bear_target,
                thesis=row.bear_thesis,
            ),
            risk_factors=row.risk_factors_json or [],
            max_drawdown_estimate=row.drawdown_estimate or "",
            position_size_pct=row.position_size_suggestion or 0.0,
            confidence=row.confidence_score or 0.0,
            timeframe=row.timeframe or "",
        )

File: backend/src/ai/reasoner.ts
/**
 * AI reasoning layer — thin orchestration between deterministic analytics
 * and the Claude API via @anthropic-ai/sdk.
 */

import Anthropic from '@anthropic-ai/sdk';
import { PrismaClient } from '@prisma/client';
import { getPrisma } from '../db/client';
import { getSettings } from '../config';
import {
  AIAnalysisResultSchema,
  ScreeningResultSchema,
  MarketOutlookSchema,
  PortfolioReviewSchema,
  type AIAnalysisResult,
  type ScreeningResult,
  type MarketOutlook,
  type PortfolioReview,
} from './schemas';
import {
  buildAnalysisPrompt,
  buildScreeningPrompt,
  buildMarketOutlookPrompt,
  buildPortfolioReviewPrompt,
  buildPortfolioQueryPrompt,
} from './prompts';

function stripCodeFences(text: string): string {
  let t = text.trim();
  if (t.startsWith('```')) {
    const lines = t.split('\n').filter((l) => !l.trim().startsWith('```'));
    t = lines.join('\n').trim();
  }
  return t;
}

function dbRowToResult(row: any): AIAnalysisResult {
  return AIAnalysisResultSchema.parse({
    bull_case: {
      probability: row.bullProbability,
      target: row.bullTarget,
      thesis: row.bullThesis,
    },
    base_case: {
      probability: row.baseProbability,
      target: row.baseTarget,
      thesis: row.baseThesis,
    },
    bear_case: {
      probability: row.bearProbability,
      target: row.bearTarget,
      thesis: row.bearThesis,
    },
    risk_factors: row.riskFactorsJson ?? [],
    max_drawdown_estimate: row.drawdownEstimate || '',
    position_size_pct: row.positionSizeSuggestion || 0,
    confidence: row.confidenceScore || 0,
    timeframe: row.timeframe || '',
  });
}

export class AIReasoner {
  private client: Anthropic;
  private settings = getSettings();

  constructor() {
    this.client = new Anthropic({ apiKey: this.settings.ANTHROPIC_API_KEY });
  }

  async analyzeTicker(
    ticker: string,
    technicalData: Record<string, any>,
    fundamentalData: Record<string, any>,
    regimeData: Record<string, any>,
    portfolioExposure: Record<string, any>,
    deep: boolean = false,
    prisma?: PrismaClient,
  ): Promise<AIAnalysisResult> {
    const db = prisma || getPrisma();
    const analysisType = deep ? 'deep' : 'standard';
    const model = deep
      ? this.settings.AI_MODEL_DEEP
      : this.settings.AI_MODEL_SCREENING;

    // Check cache
    const cutoff = new Date(
      Date.now() - this.settings.AI_CACHE_HOURS * 60 * 60 * 1000,
    );
    const cached = await db.aIAnalysis.findFirst({
      where: {
        ticker,
        analysisType,
        createdAt: { gte: cutoff },
      },
      orderBy: { createdAt: 'desc' },
    });

    if (cached) {
      console.log(`Cache hit for ${ticker} (${analysisType})`);
      return dbRowToResult(cached);
    }

    // Build prompt & call Claude
    const [systemPrompt, userPrompt] = buildAnalysisPrompt(
      ticker,
      technicalData,
      fundamentalData,
      regimeData,
      portfolioExposure,
    );

    const message = await this.client.messages.create({
      model,
      max_tokens: 2000,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
    });

    const responseText = stripCodeFences(
      message.content[0].type === 'text' ? message.content[0].text : '',
    );

    // Parse & validate
    const result = AIAnalysisResultSchema.parse(JSON.parse(responseText));

    // Persist to DB
    await db.aIAnalysis.create({
      data: {
        ticker,
        date: new Date(),
        analysisType,
        modelUsed: model,
        bullProbability: result.bull_case.probability,
        bullTarget: result.bull_case.target,
        bullThesis: result.bull_case.thesis,
        baseProbability: result.base_case.probability,
        baseTarget: result.base_case.target,
        baseThesis: result.base_case.thesis,
        bearProbability: result.bear_case.probability,
        bearTarget: result.bear_case.target,
        bearThesis: result.bear_case.thesis,
        riskFactorsJson: result.risk_factors,
        drawdownEstimate: result.max_drawdown_estimate,
        positionSizeSuggestion: result.position_size_pct,
        confidenceScore: result.confidence,
        timeframe: result.timeframe,
        createdAt: new Date(),
      },
    });

    console.log(`Saved ${analysisType} analysis for ${ticker} to DB`);
    return result;
  }

  async screenTickers(
    tickersData: Record<string, any>[],
  ): Promise<ScreeningResult[]> {
    const model = this.settings.AI_MODEL_SCREENING;

    const [systemPrompt, userPrompt] = buildScreeningPrompt(tickersData);

    const message = await this.client.messages.create({
      model,
      max_tokens: 2000,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
    });

    const responseText = stripCodeFences(
      message.content[0].type === 'text' ? message.content[0].text : '',
    );

    const rawList = JSON.parse(responseText);
    return rawList.map((item: any) => ScreeningResultSchema.parse(item));
  }

  async marketOutlook(
    regimeData: Record<string, any>,
    macroData: Record<string, any>,
    sectorData: Record<string, any>,
  ): Promise<MarketOutlook> {
    const model = this.settings.AI_MODEL_SCREENING;

    const [systemPrompt, userPrompt] = buildMarketOutlookPrompt(
      regimeData,
      macroData,
      sectorData,
    );

    const message = await this.client.messages.create({
      model,
      max_tokens: 2000,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
    });

    const responseText = stripCodeFences(
      message.content[0].type === 'text' ? message.content[0].text : '',
    );

    return MarketOutlookSchema.parse(JSON.parse(responseText));
  }

  async portfolioReview(
    positions: Record<string, any>[],
    summary: Record<string, any>,
    riskMetrics: Record<string, any>,
    healthScore: Record<string, any>,
    concentration: Record<string, any>,
    regimeData: Record<string, any>,
    force: boolean = false,
    prisma?: PrismaClient,
  ): Promise<{ review: PortfolioReview; cached: boolean; analyzed_at: string }> {
    const db = prisma || getPrisma();
    const model = this.settings.AI_MODEL_SCREENING;

    // Check cache
    if (!force) {
      const cutoff = new Date(
        Date.now() - this.settings.AI_CACHE_HOURS * 60 * 60 * 1000,
      );
      const cached = await db.aIAnalysis.findFirst({
        where: {
          ticker: 'PORTFOLIO',
          analysisType: 'portfolio_review',
          createdAt: { gte: cutoff },
        },
        orderBy: { createdAt: 'desc' },
      });

      if (cached && cached.reasoningText) {
        console.log('Cache hit for portfolio review');
        const review = PortfolioReviewSchema.parse(JSON.parse(cached.reasoningText));
        return { review, cached: true, analyzed_at: cached.createdAt.toISOString() };
      }
    }

    // Build prompt & call Claude
    const [systemPrompt, userPrompt] = buildPortfolioReviewPrompt(
      positions, summary, riskMetrics, healthScore, concentration, regimeData,
    );

    const message = await this.client.messages.create({
      model,
      max_tokens: 2000,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
    });

    const responseText = stripCodeFences(
      message.content[0].type === 'text' ? message.content[0].text : '',
    );

    const review = PortfolioReviewSchema.parse(JSON.parse(responseText));
    const now = new Date();

    // Store in DB using reasoningText as JSON blob
    await db.aIAnalysis.create({
      data: {
        ticker: 'PORTFOLIO',
        date: now,
        analysisType: 'portfolio_review',
        modelUsed: model,
        reasoningText: JSON.stringify(review),
        confidenceScore: 0,
        bullProbability: 0,
        bullTarget: '',
        bullThesis: '',
        baseProbability: 0,
        baseTarget: '',
        baseThesis: '',
        bearProbability: 0,
        bearTarget: '',
        bearThesis: '',
        riskFactorsJson: [],
        drawdownEstimate: '',
        positionSizeSuggestion: 0,
        timeframe: '',
        createdAt: now,
      },
    });

    console.log('Saved portfolio review to DB');
    return { review, cached: false, analyzed_at: now.toISOString() };
  }

  async portfolioQuery(
    query: string,
    positions: Record<string, any>[],
    summary: Record<string, any>,
    riskMetrics: Record<string, any>,
    healthScore: Record<string, any>,
    concentration: Record<string, any>,
    regimeData: Record<string, any>,
  ): Promise<string> {
    const model = this.settings.AI_MODEL_SCREENING;

    const [systemPrompt, userPrompt] = buildPortfolioQueryPrompt(
      query, positions, summary, riskMetrics, healthScore, concentration, regimeData,
    );

    const message = await this.client.messages.create({
      model,
      max_tokens: 1500,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
    });

    return message.content[0].type === 'text' ? message.content[0].text : '';
  }
}
